worker/task.go

```diff
@@ -0,0 +1,201 @@
+package worker
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/taskcluster/taskcluster-client-go/queue"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+)
+
+type TaskRun struct {
+	TaskId          string                       `json:"taskId"`
+	RunId           uint                         `json:"runId"`
+	SignedDeleteUrl string                       `json:"-"`
+	TaskClaim       queue.TaskClaimResponse      `json:"-"`
+	TaskReclaim     queue.TaskReclaimResponse    `json:"-"`
+	Definition      queue.TaskDefinitionResponse `json:"-"`
+	controller      *runtime.TaskContextController
+	context         *runtime.TaskContext
+	plugin          plugins.TaskPlugin
+	log             *logrus.Entry
+	payload         interface{}
+}
+
+func (t *TaskRun) Prepare(engine engines.Engine) (<-chan engines.SandboxBuilder, <-chan error) {
+	out := make(chan engines.SandboxBuilder, 1)
+	errc := make(chan error, 1)
+	go func() {
```

Comment:
So I see how these channels and go routines are all stitched together... Normally, you would never do this in any other language, because threads are expensive.

What is going on here is that we are starting 5 threads, where thread N waits on thread N - 1 before starting. So there is no need for the concurrency, it is all inherently serial.
Given that go routines are super cheap this is not slow or resource intensive in any concerning way.

And I won't dare to claim that channels and go routines can't be used to structure code in a convenient manner. I'm sure there are good design pattern for serial code which relies on channels and go routines. But channels and go routines are generally tools for dealing with concurrency, so maybe we should look for other ways to structure this.

I'll try to outline something here:
https://public.etherpad-mozilla.org/p/jonasfj-taskcluster-worker-task-controller